### ISSUE TITLE
The dockerng module uses deprecated API in docker-py

### ISSUE THREAD

**Issue by alf** [saltstack/salt#27661]
According to the docker-py 1.4 API documentation, providing arguments to `client.start` is decrecated. These arguments should be provided to the create method instead. This follows from a similiar change in the Docker API 1.15. We're now at 1.20 so it's high time to switch. 

A consequence of using the deprecated API is that you can't provide LogConfig arguments to docker. The reason is that the only way to provide this with docker-py is in the `host_config` provided to `client.create` however, the `host_config` given to `client.start` overrides these. 

```
Deprecation warning: For API version > 1.15, it is highly recommended to provide host config options in the host_config parameter of create_container
```

**Comment by ticosax** [comment]
Please can you try with 2015.8 branch ?

**Comment by alf** [comment]
That's the code I was reading when I noticed this issue.

https://github.com/saltstack/salt/blob/2015.8/salt/modules/dockerng.py#L421
https://github.com/saltstack/salt/blob/2015.8/salt/modules/dockerng.py#L4371

Basically, there is no such thing as a valid runtime option anymore. All options should be provided at container creation time.

**Comment by alf** [comment]
In 2015.8 in the `dockerng.running` function we split the provided arguments into runtime and create arguments. This is wrong.

https://github.com/saltstack/salt/blob/2015.8/salt/states/dockerng.py#L1428

**Comment by ticosax** [comment]
Docker-py still accepts, those arguments.
Where did you read they are deprecated ?

https://github.com/docker/docker-py/blob/master/docker/api/container.py#L248

**Comment by ticosax** [comment]
FYI #27451 has been merged one week ago.

I believe it fixes the warning you saw.

**Comment by alf** [comment]
See the [docker-py documentation](http://docker-py.readthedocs.org/en/latest/api/#start). Also see the changes between [Docker API v1.15](http://docs.docker.com/reference/api/docker_remote_api_v1.15/#start-a-container) and [Docker API v1.16](http://docs.docker.com/reference/api/docker_remote_api_v1.16/#start-a-container) where the deprecation happened.

I did not create this issue because I saw a deprecation warning. I created this issue because I'm looking forward to using the dockerng state/module and was reading through the source code to see how it looks. Since I've just been bitten by this change in the Docker API, I noticed it and wanted to give you a heads up.

**Comment by alf** [comment]
I have no doubt that the current code works perfectly fine. However the distinction between runtime and creation time arguments is no longer there so the code in dockerng can be simplified to reflect these changes.

**Comment by alf** [comment]
Also, my testing (of the Docker API) shows that host_config arguments provided at start time overrides host_config arguments provided at creation time in such a way that the arguments provided at creation time is lost.

**Comment by alf** [comment]
Notice how `Client.start` in docker-py doesn't expose the LogConfig option that was added in Docker 1.6 (API version 1.18). If you want to provide this you'll have to provide it in the host_config argument to `Client.create_container`. Unfortunately, since dockerng provides host_config arguments to `Client.start` the LogConfig argument gets dropped.

**Comment by ticosax** [comment]
Thank you @alf. I see what you mean now. I didn't realize the API changed so much. I will come up with a patch and some tests.

**Comment by alf** [comment]
Thank you. The API changes surprised me too. I do prefer the new way however since it's much less confusing what goes where. :)

**Comment by ticosax** [comment]
>  I do prefer the new way however since it's much less confusing what goes where.

Sure. The difficulty is to keep backward compatibility with older docker daemon. That's why I got bitten by api of docker-py.

**Comment by alf** [comment]
Yes, I can see how that complicates things. Thanks again for the good work. I really like how dockerng is turning out.

**Comment by ticosax** [comment]
@terminalmage is behind dockerng. I'm just a user who fixes bugs here and there.

**Comment by ticosax** [comment]
I've been thinking: Since the new API has been introduced docker api version 1.16
We can certainly drop API usage of `start` with parameters, completely.

This API has been released with docker 1.4.0 in December 2014.
https://github.com/docker/docker/commit/ca6cc6560987a87b8f3b4fe4c3c227d692c49117
I'd say it is long enough to drop support of API versions prior of `1.16`

@terminalmage What do you think ?

**Comment by alf** [comment]
What's the deprecation timeline for the dockerio module? That's probably relevant for this discussion.

**Comment by ticosax** [comment]
I made a POC Pull request. subject to change.

@alf please can you give a try ?

**Comment by terminalmage** [comment]
@alf We're going to keep both modules side by side for at least a couple releases, dockerio.py won't go away for a while still.

**Comment by alf** [comment]
In that case I suggest that dockerng drops backwards compatibility with docker API 1.15. That means dockerng requires docker >= 1.4.

People who uses older versions of docker can use dockerio.

**Comment by alf** [comment]
@ticosax I'll try to test your changes tomorrow.

**Comment by ticosax** [comment]
@alf nice to hear !
I'll wrote more test for it once I have time.

**Comment by alf** [comment]
Sorry for not getting around to testing this. Life and work became very hectic all of a sudden. Happy to see it pass through anyway.

**Comment by ticosax** [comment]
@jfindlay This issue can be closed since #27676 has been merged.
